cmake/DLAF_AddTest.cmake

```
@@ -8,21 +8,125 @@
 # SPDX-License-Identifier: BSD-3-Clause
 #
 
-# DLAF_addTest(test_target_name
-#   SOURCES <source1> [<source2> ...]
-#   [COMPILE_DEFINITIONS <arguments for target_compile_definitions>]
-#   [INCLUDE_DIRS <arguments for target_include_directories>]
-#   [LIBRARIES <arguments for target_link_libraries>]
+set(DLAF_PIKATEST_EXTRA_ARGS "" CACHE STRING "Extra arguments for tests with pika")
+
+macro(dlaf_setup_mpi_preset)
+  # if a preset has been selected and it has been changed from previous configurations
+  if(NOT DLAF_MPI_PRESET STREQUAL _DLAF_MPI_PRESET)
+
+    if(DLAF_MPI_PRESET STREQUAL "plain-mpi")
+      # set(MPIEXEC_EXECUTABLE "" CACHE STRING "Executable for running MPI programs")
+      # set(MPIEXEC_NUMPROC_FLAG "" CACHE STRING "Flag used by MPI to specify the number of processes for mpiexec; the next option will be the number of processes." FORCE)
+      set(MPIEXEC_NUMCORE_FLAG
+          ""
+          CACHE
+            STRING
+            "Flag used by MPI to specify the number of cores per rank for mpiexec. If not empty, you have to specify also the number of cores available per node in MPIEXEC_NUMCORES_PER_RANK."
+            FORCE
+      )
+      set(MPIEXEC_NUMCORES_PER_RANK "" CACHE STRING "Number of cores used by each MPI rank." FORCE)
+      if(DLAF_TEST_THREAD_BINDING_ENABLED)
+        message(WARNING "Disabling pika binding")
+        set(DLAF_TEST_THREAD_BINDING_ENABLED FALSE CACHE BOOL "" FORCE)
+      endif()
+
+    elseif(DLAF_MPI_PRESET STREQUAL "slurm")
+      if(NOT DLAF_TEST_THREAD_BINDING_ENABLED)
+        message(
+          WARNING "When using DLAF_MPI_PRESET=slurm DLAF_TEST_THREAD_BINDING_ENABLED should be enabled. "
+                  "It is currently disabled and you may incur in performance drop. "
+                  "Leave it disabled at your own risk."
+        )
+      endif()
+
+      execute_process(
+        COMMAND which srun OUTPUT_VARIABLE SLURM_EXECUTABLE OUTPUT_STRIP_TRAILING_WHITESPACE
+      )
+      set(MPIEXEC_EXECUTABLE ${SLURM_EXECUTABLE} CACHE STRING "Executable for running MPI programs"
+                                                       FORCE
+      )
+      set(MPIEXEC_NUMPROC_FLAG
+          "-n"
+          CACHE
+            STRING
+            "Flag used by MPI to specify the number of processes for mpiexec; the next option will be the number of processes."
+            FORCE
+      )
+      set(MPIEXEC_NUMCORE_FLAG
+          "-c"
+          CACHE
+            STRING
+            "Flag used by MPI to specify the number of cores per rank for mpiexec. If not empty, you have to specify also the number of cores available per node in MPIEXEC_NUMCORES_PER_RANK."
```

```suggestion
            "Flag used by MPI to specify the number of cores per rank for mpiexec. If not empty, you have to specify also the number of cores available per rank in MPIEXEC_NUMCORES_PER_RANK."
```
as in the other docstring. But also, any way we can combine the two definitions?

---

It's a good point that I totally agree with. Unfortunately using CMake CACHE variables means that we have to `FORCE` set them if we want to change the value. But changing the value also implies setting also the related docstring (AFAIK).

I think the only way might be storing docstring in variables. Do you see any alternative? would you go for variables in order to not be error-prone and risking ending up with same variable with different docstring?